### DISCUSSION TITLE
Non-static TestCaseSource/ValueSource

### DISCUSSION THREAD

**Discussion by manfred-brands** [nunit/nunit#5071] [Design]
At the moment we require the members that are passed into the `TestCaseSource` and `ValueSource` attributes to be `static`.

However, if the name is `null` we assume the `TestFixture` itself is an `IEnumerable` and we actually instantiate the fixture to return the test parameters.

What is the reason, not to allow _instance_ fields for the other source members?
Code wise the change wouldn't be too difficult.

`object instance = member.IsStatic ? null : Reflect.Construct(sourceType, null)`

Then update all further code to pass `instance` instead of `null`?

**Comment by stevenaw** [comment]
@manfred-brands I found a bit of a write up about the history of this here: https://github.com/nunit/nunit/issues/3532#issuecomment-626189448

That comment mentions avoiding multiple construction of objects. My (perhaps incorrect) assumption had been that it also could encourage more deterministic behavior by avoiding referencing instance data when generating test cases.

I admit, I haven't read the whole thread just yet. And, of course, decisions can always be revisited. 

@jnm2 At one point you were advocating for going further towards `static`s in https://github.com/nunit/nunit/issues/3537. I'm unsure if that was simply for API consistency or if you had other design considerations in mind too at the time?
